src/core/vector/qgsvectorlayerprofilegenerator.cpp

```diff
@@ -72,6 +72,165 @@ QVector<QgsGeometry> QgsVectorLayerProfileResults::asGeometries() const
   return res;
 }
 
+QgsAbstractProfileResults::SnapResult QgsVectorLayerProfileResults::snapPoint( double distanceAlongCurve, double height, double maximumCurveDelta, double maximumHeightDelta )
+{
+  // TODO -- add spatial index if performance is an issue
+  QgsAbstractProfileResults::SnapResult res;
+
+  for ( auto it = features.constBegin(); it != features.constEnd(); ++it )
+  {
+    for ( const Feature &feature : it.value() )
+    {
+      const QgsRectangle featureBounds = feature.crossSectionGeometry.boundingBox();
+      if ( ( featureBounds.xMinimum() - maximumCurveDelta <= distanceAlongCurve ) && ( featureBounds.xMaximum() + maximumCurveDelta >= distanceAlongCurve ) )
+      {
+        switch ( feature.crossSectionGeometry.type() )
+        {
+          case QgsWkbTypes::PointGeometry:
+          {
+            for ( auto partIt = feature.crossSectionGeometry.const_parts_begin(); partIt != feature.crossSectionGeometry.const_parts_end(); ++partIt )
+            {
+              if ( const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( *partIt ) )
+              {
+                const double snapDistanceDelta = std::fabs( distanceAlongCurve - point->x() );
+                if ( snapDistanceDelta > maximumCurveDelta )
+                  continue;
+
+                const double snapHeightDelta = std::fabs( height - point->y() );
+                if ( snapHeightDelta > maximumHeightDelta )
+                  continue;
+
+                const double snapDistance = point->distance( QgsPoint( distanceAlongCurve, height ) );
```

Comment:
Do I read it correctly that snapDistance is using horizontal+vertical distance for vectors, but only using vertical distance for raster and mesh?

I am wondering how we should be treating these distances - given that X/Y axes in profile plot may have different scales, shall that be reflected in the snap distance calculation somehow? Maybe keep horizontal and vertical snap distances separately, and when calculating the "total" snap distance, the X/Y axis ratio would be involved as well - or am I overcomplicating this? :-)

Reply:
> Do I read it correctly that snapDistance is using horizontal+vertical distance for vectors, but only using vertical distance for raster and mesh?

Yes, you're correct. I originally had horizontal snapping for raster/mesh too, but it just felt "wrong" in practice. E.g. in this situation:

```
        o       /------------
            /(x)
-------+/
```

it would result in the point horizontal snapping to "X", whereas in practice it felt much nicer when the snap was vertically constrained for these continuous surfaces, resulting in snapping to "+".

So then I disabled horizontal snapping completely, but that meant you couldn't snap to points/short horizontal lines anymore => using horizontal snapping only on vectors resulted in best user experience (in my opinion, at least)!. I'll also enable horizontal snap for point cloud points for the same reason.

> I am wondering how we should be treating these distances - given that X/Y axes in profile plot may have different scales, shall that be reflected in the snap distance calculation somehow? Maybe keep horizontal and vertical snap distances separately, and when calculating the "total" snap distance, the X/Y axis ratio would be involved as well - or am I overcomplicating this? :-)

I think (given the comment above) I'll rework the "total snapped distance" member and remove that, and calculate that only during collation of the different source snapping results (since we can easily calculate the horizontal/vertical displacement from the original point there). I'll also add a "snap context" argument so that we've got a way to specify the horizontal/vertical ratio display ratio too.